gtime/causality/granger_causality.py

```diff
@@ -0,0 +1,213 @@
+    >>> import pandas.util.testing as testing
+    >>> data = testing.makeTimeDataFrame(freq="s", nper=1000)
+    >>> GrangerCausality(target_col='A', x_col='B', max_shift=10, statistics='ssr_F').fit(data)
+                        ssr F-test
+    F-value               0.372640
+    p-value               0.958527
+    degrees of freedom  969.000000
+    number of shifts     10.000000
+    """
+    def __init__(self, 
+                 target_col: str, 
+                 x_col: str, 
+                 max_shift: int, 
+                 statistics='ssr_F'):
+        self.target_col = target_col
+        self.x_col = x_col
+        self.max_shift = max_shift
+        self.statistics = statistics
+
+    def fit(self, data: pd.DataFrame):
+        """Create a dataframe with the results of the Granger causality test with the specified
+        statistical test(s).
+
+        Parameters
+        ----------
+        data : pd.DataFrame, shape (n_samples, n_time_series), required
+            The dataframe containing the time series.
+
+        Returns
+        -------
+        result : pd.DataFrame (if only one test is selected) or tuple if statics=='all'
+            A dataframe or a tuple of dataframes for each selected statistical test.
+        """
+        x = self.x_col
+        y = self.target_col
+
+        shifts = pd.DataFrame()
+        
+        for shift in range(0, self.max_shift+1):
+            shifts[str(x)+'_'+str(shift)] = data[x].shift(shift)
+            if shift > 0:
+                shifts[str(y)+'_'+str(shift-1)] = data[y].shift(shift)
+        shifts = shifts.dropna()
+                
+        y_columns = [c for c in shifts.columns if c.split('_')[0]==y]
+        x_columns = [c for c in shifts.columns if c.split('_')[0]==x]
+        
+        data_single = shifts[x_columns].copy()
+        data_joint = shifts[x_columns + y_columns].copy()
+        data_single.drop([str(x)+'_'+str(0)], inplace=True, axis='columns')
+        data_joint.drop([str(x)+'_'+str(0)], inplace=True, axis='columns')
+
+        lr_single = LinearRegression()
+        lr_joint = LinearRegression()
+        lr_single.fit(data_single, data[x].loc[data_single.index])
+        lr_joint.fit(data_joint, data[x].loc[data_joint.index])
+
+        y_pred_single = lr_single.predict(data_single)
+        y_pred_joint = lr_joint.predict(data_joint)
+        
+        try:
+            DOF_single = float(data_single.shape[0] - data_single.shape[1]) 
+        except:
+            DOF_single = float(data_single.shape[0] - 1)
+        DOF_joint = float(data_joint.shape[0] - data_joint.shape[1]) -1  
+                
+        lr_single_residues = lr_single._residues
+        lr_joint_residues = lr_joint._residues
+
+        result = []
+        if self.statistics=='all':
+            stat = ['ssr_F', 'ssr_chi2', 'likelihood_chi2', 'zero_F']
+        else:
+            stat = [self.statistics]
+
+        if 'ssr_F' in stat:
```

Comment:
Personally, I would split all these if-blocks in smaller methods. This would improve readability and, if needed, to unittest smaller parts of the code.